### DISCUSSION TITLE
Measuring the last row of RichText during word wrapping

### DISCUSSION THREAD

**Discussion by hkparker** [fyne-io/fyne#5298] [Q&A] [answered]
Is it possible to determine what the last row of wrapped RichText is via the API, such that I can `fyne.MeasureText` just that part of it?  I'm trying to neatly pack some widgets into a chat bubble, and when I'm wrapping it would be quite useful to know how much space I'm using on that last line so I can check if I have room to put icons there without making the bubble bigger.

**Comment by hkparker** [comment]
I tried to copy [`lineBounds`](https://github.com/fyne-io/fyne/blob/v2.5.2/widget/richtext.go#L944) out, however quickly ran into internal libraries (paint) that I couldn't pull out quite so easily

**Comment by andydotxyz** [comment]
You can (and may have to) query the underlying go-text library as that information is returned by the shaping run.

I don't even know how we would expose this as wrapping is highly context sensitive so any query to this state would somehow have to go through a text widget's renderer I think..?

**Comment by hkparker** [comment]
Yeah I'm doing work in a custom renderer right now.  I wouldn't suggest it makes sense to actually expose this in the API, it's a weird request, but it's something I need to make my app look as good as it can, so I'm curious to see what I can do.  Probably that means re-computing the word wrapping an extra time in my code just for this (if that doesn't end up being too expensive).

I see the shaping library used down in [`truncateLimit`](https://github.com/fyne-io/fyne/blob/v2.5.2/widget/richtext.go#L1114) but it looks like what I'm interested in would be [this](https://github.com/fyne-io/fyne/blob/v2.5.2/widget/richtext.go#L1006), which doesn't use it, so I'm not sure.

If I do pursue this I think I might do something like split the last line into chunks of text plus their leading whitespace

```
"this is a message    with some spaces" => ["this", " is", " a", " message", "    with", " some", " spaces"]
```

then measure and store the length of each segment with `fyne.MeasureText`.  Then when I want to see what the last line would be and how much it would take up, I can pop segments off into groups that just fit into my target width until I'm left with the last line, and sum all the segment lengths there.

That'd be a lot of work!  But it should be accurate.

I tried a jankier `availableWidth - mat.Mod(line, availableWidth)` which assumes perfectly even (not word) wrapping, and that was just off enough to not work.

**Comment by andydotxyz** [comment, marked as answer]
Opened a feature request from this info https://github.com/fyne-io/fyne/issues/5700
